sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs

```diff
@@ -58,19 +54,114 @@ private enum SupportedType
+        #region Encode
+
+        /// <summary>
+        /// Encodes the message data into Avro and stores it in <see cref="MessageWithMetadata.Data"/>. The <see cref="MessageWithMetadata.ContentType"/>
+        /// will be set to "avro/binary+schemaId" where schemaId is the ID of the schema used to encode the data.
+        /// </summary>
+        /// <param name="data">The data to serialize to Avro and encode into the message.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        /// <typeparam name="TMessage">The <see cref="MessageWithMetadata"/> type to encode the data into.</typeparam>
+        /// <typeparam name="TData">The type of the data to encode.</typeparam>
+        public TMessage EncodeMessageData<TMessage, TData>(
+            TData data,
+            CancellationToken cancellationToken = default) where TMessage : MessageWithMetadata, new()
+            => (TMessage) EncodeMessageDataInternalAsync(data, typeof(TData), typeof(TMessage), false, cancellationToken).EnsureCompleted();
+
+        /// <summary>
+        /// Encodes the message data into Avro and stores it in <see cref="MessageWithMetadata.Data"/>. The <see cref="MessageWithMetadata.ContentType"/>
+        /// will be set to "avro/binary+schemaId" where schemaId is the ID of the schema used to encode the data.
+        /// </summary>
+        /// <param name="data">The data to serialize to Avro and encode into the message.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        /// <typeparam name="TMessage">The <see cref="MessageWithMetadata"/> type to encode the data into.</typeparam>
+        /// <typeparam name="TData">The type of the data to encode.</typeparam>
+        public async ValueTask<TMessage> EncodeMessageDataAsync<TMessage, TData>(
+            TData data,
+            CancellationToken cancellationToken = default) where TMessage : MessageWithMetadata, new()
+            => (TMessage) await EncodeMessageDataInternalAsync(data, typeof(TData), typeof(TMessage), true, cancellationToken).ConfigureAwait(false);
+
+        /// <summary>
+        /// Encodes the message data into Avro and stores it in <see cref="MessageWithMetadata.Data"/>. The <see cref="MessageWithMetadata.ContentType"/>
+        /// will be set to "avro/binary+schemaId" where schemaId is the ID of the schema used to encode the data.
+        /// </summary>
+        /// <param name="data">The data to serialize to Avro and encode into the message.</param>
+        /// <param name="dataType">The type of the data to encode. If left blank, the type will be determined at runtime by
+        /// calling <see cref="Object.GetType"/>.</param>
+        /// <param name="messageType">The type of message to encode the data into. Must extend from <see cref="MessageWithMetadata"/>.
```

Comment:
Should we call out the need for a parameterless constructor?

```suggestion
        /// <param name="messageType">The type of message to encode the data as. Must extend from <see cref="MessageWithMetadata"/> and have a parameterless constructor.
```

Reply:
It isn't really encoding the data as the type, but rather into the Data property of the type.